src/hotspot/os/linux/os_linux.cpp

```diff
@@ -6019,8 +6021,143 @@ class CracSHM {
   }
 };
 
+class Freeze {
+  static const int signo = SIGUSR1;
+  struct sigaction act_old;
+
+  template<class Callback> bool all_threads(Callback callback) const {
+    const char dirname[] = "/proc/self/task";
+    GrowableArray<pid_t> done;
+    bool signalled;
+    do {
+      signalled = false;
+      DIR *dir = opendir(dirname);
+      if (dir == NULL) {
+        tty->print_cr("JVM: Error opening %s: %m", dirname);
+        return false;
+      }
+      pid_t tid = gettid();
+      struct dirent *ent;
+      while (errno = 0, ent = readdir(dir)) {
+        const char *name = ent->d_name;
+        if (strcmp(name, ".") == 0)
+          continue;
+        if (strcmp(name, "..") == 0)
+          continue;
+        char *endptr;
+        unsigned long l = strtoul(name, &endptr, 10);
+        pid_t ent_tid = l;
+        if (l >= LONG_MAX || ent_tid != (long) l) {
```

Comment:
`l` is an `unsigned long` and I think `LONG_MAX` is a signed value as there's also a `ULONG_MAX`.  Are we using this to validate that `l` is a positive value?  Otherwise, I'm not sure I understand this condition.

Reply:
The number must fit in `pid_t` (otherwise it is a weird Linux kernel). Therefore `int`. But there is no `strtoi`. Also PID must not be negative (otherwise it is a weird Linux kernel again). So I found this as the most simple and safe code to validate `name` contains a non-negative number which can fit in `pid_t`.

Reply:
> The number must fit in `pid_t` (otherwise it is a weird Linux kernel). Therefore `int`. But there is no `strtoi`.

If we're boxing to an int, then I think we should check if `l >= INT_MAX` rather than `LONG_MAX`.

Reply:
The `int` check is there `ent_tid != (long) l` (as nobody says `pid_t` is really an `int`). The `l >= LONG_MAX` is there to make valid the cast `(long) l`. But this is IMO a [bikeshed](https://en.wikipedia.org/wiki/Law_of_triviality).